PD-129 added cases to handle filter and sort for json fields.

anubhav0fnu opened the pull request:
Ticket: https://teamheadlight.atlassian.net/browse/PD-129
Approach: Try prototyping SQL queries with [Foxhound](https://github.com/stevenvelozo/foxhound). It works we need to add support to the meadow filter.
 ```
 var _Fable = require('fable').new();
 var libFoxHound = require('foxhound');
 
var tmpQuery = libFoxHound
.new(_Fable)
.setDialect('MySQL')
.setScope('Document')
.addFilter('length(FormDataJson)', '1', '>', 'AND', 'FormDataJson_length')
.addFilter("JSON_EXTRACT(FormDataJson, '$.Header.SampleMaterial')", 'Profile Testing', '=', 'AND','jsonFieldKey');

  console.log(tmpQuery.buildReadQuery().query.body);
  // ----
var tmpQuery = libFoxHound
 .new(_Fable)
 .setDialect('MySQL')
 .setScope('Document')
 .setDataElements(['IDDocument', 'CreateDate', 'FormDataJson', "FormDataJson->'$.Header.SampleMaterial'"])
 .addFilter('length(FormDataJson)', '1', '>', 'AND', 'FormDataJson_length')
 .setSort([{ Column: "FormDataJson->'$.Header.SampleMaterial'", Direction: 'Descending' }]);

console.log(tmpQuery.buildReadQuery().query.body);
```
generates:
```
   SELECT `IDDocument`, `CreateDate`, `FormDataJson` FROM `Document` WHERE length(FormDataJson) > :FormDataJson_length_w0 AND JSON_EXTRACT(FormDataJson, '$.Header.SampleMaterial') = :jsonFieldKey_w1;
   ---
   SELECT `IDDocument`, `CreateDate`, `FormDataJson`, `FormDataJson->'$`.`Header` FROM `Document` WHERE length(FormDataJson) > :FormDataJson_length_w0 ORDER BY FormDataJson->'$.Header.SampleMaterial' DESC;
```
After which added the case stmts to support new filter and sort operations in json fields.

--
New changes:

https://localhost/1.0/Documents/FilteredTo/FSJF~FormDataJson,Header.SampleMaterial~DESC~0
```
SELECT `Document`.* 
FROM `Document` 
WHERE  LENGTH(FormDataJson) > "1" AND 
`Document`.Deleted = 0 
ORDER BY FormDataJson->'$.Header.SampleMaterial' DESC, DocumentDate DESC LIMIT 250;
```
https://localhost/1.0/Documents/FilteredTo/FBJV~FormDataJson,Header.SampleMaterial~EQ~Profile Testing
```
SELECT `Document`.* 
FROM `Document` 
WHERE  LENGTH(FormDataJson) > "1" AND 
JSON_EXTRACT(FormDataJson, '$.Header.SampleMaterial') = "Profile Testing" AND  
`Document`.Deleted = 0 
ORDER BY DocumentDate DESC LIMIT 250;
```
Steps: 
to repro locally, 

1. copy the latest changes of this repo to mpi_api `node_modules`
   `cp -r /Pavia/meadow-filter/source/* /Pavia/mpi_api/node_modules/meadow-filter/source`
2. and touch any file in mpi_api may be`/Pavia/mpi_api/ts_source/bundles/contentmanagement/Documents.ts`, so that changes are picked up by the watcher.
3. check query in docker logs
   `docker logs pavia-headlight-api-1 -n 100 --follow`
   
 Tests:
  check: sort the filtered documents and then return the pagination results back: `0th matches 10th record`
 ```
https://localhost/1.0/Documents/FilteredTo/FBJV~FormDataJson,Header.SampleMaterial~EQ~Profile Testing~FSJF~FormDataJson,Header.SampleMaterial~DESC~0/10/10
https://localhost/1.0/Documents/FilteredTo/FBJV~FormDataJson,Header.SampleMaterial~EQ~Profile Testing~FSJF~FormDataJson,Header.SampleMaterial~DESC~0/0/20
 ```

anubhav0fnu committed:
BE-708 added cases to handle filter and sort for json fields.

anubhav0fnu committed:
PD-129
addded correct parsing logic to support jsonfields.

anubhav0fnu committed:
PD-129- changes to filters are working as desired now.

bigal-hl commented on source/Meadow-Filter-Parse.js:
```diff
@@ -160,6 +160,33 @@ const addFilterStanzaToQuery = (pFilterStanza, pQuery) =>
+      {
+      const tmpSortDirection = pFilterStanza.Operator === 'DESC' ? 'Descending' : 'Ascending';
+      const jsonField = pFilterStanza.Field.split(',')[0];
+      const subFields = pFilterStanza.Field.split(',')[1];
```
There is no reason to split this twice. Also, this should validate the input and not generating something invalid below (ex. if there isn't a comma). Also, this breaks if there are > 1 comma - I would use the two parameter version of split to address.

(same thing above)

anubhav0fnu replied:
Yes, if there is  >1 comma, no comma, or some other delimiter, it errors out. Spitting twice taken care off.

bigal-hl commented on source/Meadow-Filter-Parse.js:
```diff
@@ -160,6 +160,33 @@ const addFilterStanzaToQuery = (pFilterStanza, pQuery) =>
+      const jsonField = pFilterStanza.Field.split(',')[0];
+      const subFields = pFilterStanza.Field.split(',')[1];
+      pQuery
+        .addFilter('', '1', `LENGTH(${jsonField}) >`, 'AND', 'jsonFieldLen')
```
This reads weird - why are these fields out of order? Is this a hack or workaround of some kind?

anubhav0fnu replied:
it's a workaround. If I use `.addFilter('length(FormDataJson)', '1', '>', 'AND', 'FormDataJson_length')` the foxhound is adding Document.LENGTH(FormDataJson)

bigal-hl replied:
I thought we agreed that the right fix for that was prefixing the field with the table name when adding it?

Did we track down where that behavior was coming from? Is there a reason this solution won't work?

anubhav0fnu committed:
PD-129- addressed comments.

jy-headlight reviewed (commented):
Overall, I'm not a meadow expert, though the sql you mentioned seems valid. I'll try your repro steps. Alex also mentioned some fields were in an unexpected order for your workaround. 

- Prefer that testing is added, though maybe can add later?: add a short unit test or two in https://github.com/bigal-hl/meadow-filter/blob/main/test/test.js 
- nits: spacing is in consistent with the rest of the code. CMD + P, and Format Document in VS Code should fix up a lot for you. Or highlight a section and something like Format Selection should also be available under CMD + P.
- The ACs of the ticket mention adding some benchmarking data. Maybe you can list some speeds locally, and when it's on QA mention what speeds you're finding there.

- I'm gonna guess the client will pass in an IDProject in the dashboard usecase at least, so at least the perf hit in practice should be better.

anubhav0fnu committed:
PD-129- addressed comments: validation of the input

jy-headlight commented:
Tried repro steps and since I cannot upload json payloads here (github doesn't allow?) I added comments on your ticket. The functionality I can confirm works. 

I played with the filters and decided to sort by a different json property than what's in the `FSJF` filter. There's a slight quirk in that `JCSequenceNumber` `"9"` will appear before `"88"`. This is from javascript's lexicographic sorting when it comes to strings. I think the form data, almost all values are strings. Not sure if we need to consider "fixing" this, or just to document it well. --I think this would be out of scope for the ticket though as the ticket mentions fields that are not numeric, so let's punt on this.

The endpoint I tried:

https://localhost/1.0/Documents/FilteredTo/FBJV~FormDataJson,Header.SampleMaterial~EQ~Profile%20Testing~FSJF~FormDataJson,Header.JCSequenceNumber~DESC~0/10/10

kamaurra-headlight commented on source/Meadow-Filter-Parse.js:
```diff
@@ -160,6 +160,47 @@ const addFilterStanzaToQuery = (pFilterStanza, pQuery) =>
+      const fields = pFilterStanza.Field.split(',', 2);
+      const jsonField =fields[0];
+      const subFields = fields[1];
+      if (jsonField !== undefined && jsonField.trim() !== "" && subFields !== undefined && subFields.trim() !== "")
```
nit: flip the if and else

anubhav0fnu committed:
PD-129 -
* added integration test:
run it using
  npm run test:integration or
  mocha -u tdd --timeout 20000 --recursive test/integration/jsonFilterTest.js

* made changes in package.json

bigal-hl commented on source/Meadow-Filter-Parse.js:
```diff
@@ -160,6 +160,47 @@ const addFilterStanzaToQuery = (pFilterStanza, pQuery) =>
 			}
 			break;
 
+    case 'FBJV':// Filter by JSON Value (left-side AND)
+      {
+      const fields = pFilterStanza.Field.split(',', 2);
+      const jsonField =fields[0];
+      const subFields = fields[1];
+      if (jsonField !== undefined && jsonField.trim() !== "" && subFields !== undefined && subFields.trim() !== "")
+        {
+      pQuery
+        .addFilter('', '1', `LENGTH(${jsonField}) >`, 'AND', 'jsonFieldLen')
```
Inconsistent indent (tabs vs. spaces).

bigal-hl commented on source/Meadow-Filter-Parse.js:
```diff
@@ -160,6 +160,47 @@ const addFilterStanzaToQuery = (pFilterStanza, pQuery) =>
+      const fields = pFilterStanza.Field.split(',', 2);
+      const jsonField =fields[0];
+      const subFields = fields[1];
+      if (jsonField !== undefined && jsonField.trim() !== "" && subFields !== undefined && subFields.trim() !== "")
```
(nit) Just doing a truthy check seems simpler here, since `''` is falsy. See: https://developer.mozilla.org/en-US/docs/Glossary/Falsy

bigal-hl commented on source/Meadow-Filter-Parse.js:
```diff
@@ -160,6 +160,47 @@ const addFilterStanzaToQuery = (pFilterStanza, pQuery) =>
+      {
+      const tmpSortDirection = pFilterStanza.Operator === 'DESC' ? 'Descending' : 'Ascending';
+      const fields = pFilterStanza.Field.split(',', 2);
+      const jsonField =fields[0];
+      const subFields = fields[1];
+      if (jsonField !== undefined && jsonField.trim() !== "" && subFields !== undefined && subFields.trim() !== "")
+        {
```
It's weird that we are trimming here but not for the actual value used. If we need that, it feels better to do all of this during the extraction. ex.

```javascript
const jsonField = (fields[0] || '').trim();
const subFields = (fields[1] || '').trim();
```

Also (nit) generally, I would prefer to write this as a guard clause, where we bail out if we can't continue (reduces indentation).

kamaurra-headlight commented on package.json:
```diff
@@ -5,7 +5,8 @@
     "coverage": "nyc npm run test && nyc report --reporter=lcov",
-    "test": "./node_modules/.bin/mocha --exit -u tdd -R spec"
+    "test": "mocha --ignore \"test/integration/*.js\" --exit -u tdd --recursive",
+    "test:integration": "mocha -u tdd --timeout 20000 --recursive test/integration/jsonFilterTest.js"
```
nit: might not want to limit to a single file
```suggestion
    "test:integration": "mocha -u tdd --timeout 20000 --recursive test/integration/*.js"
```

bigal-hl commented on source/Meadow-Filter-Parse.js:
```diff
@@ -160,6 +160,47 @@ const addFilterStanzaToQuery = (pFilterStanza, pQuery) =>
 			}
 			break;
 
+    case 'FBJV':// Filter by JSON Value (left-side AND)
```
Any reason we aren't supporting:
* `FBJVOR`
* `FBJL`
* `FBJLOR`

kamaurra-headlight replied:
nit: this is a personal preference for `FBVJ`, so it lines up neater with the existing `FBV`.

bigal-hl commented on source/Meadow-Filter-Parse.js:
```diff
@@ -160,6 +160,47 @@ const addFilterStanzaToQuery = (pFilterStanza, pQuery) =>
+      break;
+      }
+
+    case 'FSJF':// Filter Sort Field
```
Comment mismatch.

kamaurra-headlight commented on test/integration/jsonFilterTest.js:
this integration should go in the API service. 🤔 It's actually pinging the API service. We don't want tests in this small library to fail because the headlight ecosystem is not available or changes.

bigal-hl commented on source/Meadow-Filter-Parse.js:
```diff
@@ -160,6 +160,47 @@ const addFilterStanzaToQuery = (pFilterStanza, pQuery) =>
 			}
 			break;
 
```
Missing documentation for new stanzas here: https://github.com/bigal-hl/meadow-filter/blob/main/source/Meadow-Filter-Parse.js#L45

kamaurra-headlight commented on test/integration/jsonFilterTest.js:
```diff
@@ -0,0 +1,27 @@
+
+suite('Method test', async () =>
+{
+  test('Check Sort', async () => {
```
I'm not sure this test is doing what it set out to do. It's validating that **some data is returned**, but it should be checking that the **right data is returned**.

kamaurra-headlight commented on test/integration/jsonFilterTest.js:
```diff
@@ -0,0 +1,27 @@
+		const tenToTwentyResult = await api.get(tenToTwentyQuery);
+		expect(tenToTwentyResult.body[0]['IDDocument']).to.deep.equal(firstTwentyResult.body[10]['IDDocument']);
+	});
+});
```
nit: newline at EOF
```suggestion
});

```

bigal-hl commented on source/Meadow-Filter-Parse.js:
```diff
@@ -160,6 +160,47 @@ const addFilterStanzaToQuery = (pFilterStanza, pQuery) =>
+        {
+      pQuery
+        .addFilter('', '1', `LENGTH(${jsonField}) >`, 'AND', 'jsonFieldLen')
+        .addFilter(
```
Since we are adding two clauses here in one, I think we need to add parentheses, otherwise the input query may be broken.

kamaurra-headlight commented on test/integration/jsonFilterTest.js:
```diff
@@ -0,0 +1,27 @@
+suite('Method test', async () =>
+{
+  test('Check Sort', async () => {
+    // check: sort the filtered documents and then return the pagination results back: 0th matches 10th record
```
nit: indent

bigal-hl commented on test/integration/client.js:
```diff
@@ -0,0 +1,58 @@
+		https: {
+			rejectUnauthorized: false,
+		},
+		hooks: {
+			timeout: 15000,
+			beforeRequest: [
+				(options) => {
+					const postBody = `${
+						options.method === 'POST' ? `: ${JSON.stringify(options.json)}` : ''
+					}`;
+					console.info(
+						`Going to send ${options.method} to API on URL [${options.url}]${postBody}`
+					);
+				},
+			],
+			afterResponse:
+			[
+				async (response, retryWithMergedOptions) =>
+				{
+					if (response.body && response.body.Error)
+					{
```
(nit) Inconsistent brace style.

bigal-hl commented on test/integration/jsonFilterTest.js:
```diff
@@ -0,0 +1,27 @@
+		const firstTwentyQuery = `Documents/FilteredTo/${fbjvFilter}~${fsjfFilter}/0/20`;
+		const tenToTwentyQuery = `Documents/FilteredTo/${fbjvFilter}~${fsjfFilter}/10/10`;
+
+		const firstTwentyResult = await api.get(firstTwentyQuery);
+		const tenToTwentyResult = await api.get(tenToTwentyQuery);
```
(nit) Could parallelize.

bigal-hl commented on test/integration/jsonFilterTest.js:
```diff
@@ -0,0 +1,27 @@
+
+suite('Method test', async () =>
+{
+  test('Check Sort', async () => {
+    // check: sort the filtered documents and then return the pagination results back: 0th matches 10th record
+		await api.post('Authenticate', { json: { UserName: apiSettings.UserName, Password: apiSettings.Password } });
```
(nit) Inconsistent indent.

bigal-hl commented on package.json:
```diff
@@ -22,10 +23,13 @@
     "mocha": "^9.2.2",
     "nyc": "^15.1.0",
     "sinon": "^13.0.1"
+  },
+  "dependencies": {
+    "@paviasystems/service_pieces": "^1.0.48"
   }
```
No way should this be a dependency. It could be a dev dependency.

kamaurra-headlight reviewed (changes requested):
I believe we discuss putting unit tests in here, and integration tests potentially in the API service.
I also feel the one test written doesn't achieve enough.
Other than that, mostly nitpicks. I don't understand the workaround in place for the `.addFilter` call.

kamaurra-headlight commented:
closing in favor of #2
